[CELEBORN-1894] Allow skipping already read chunks during unreplicated shuffle read retried

saurabhd336 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  - Make sure the PR title start w/ a JIRA ticket, e.g. '[CELEBORN-XXXX] Your PR title ...'.
  - Be sure to keep the PR description updated to reflect all changes.
  - Please write your PR title to summarize what this PR proposes.
  - If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
Whenever a `WorkerPartitionReader` is recreated (due celeborn worker restarts / any other chunk fetch failure), the entire shuffle partition file is re-read from beginning, discarding already read chunks in `CelebornInputStream` based on the batchIdSet metadata maintained.

This can be improved (only for cases where shuffle data is unreplicated) by skipping already read chunk id since they'd be discarded anyway. This improves overall shuffle read performance (reducer's total time, network usage etc).

### Why are the changes needed?
Allow skipping already read shuffle chunks 


### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UTs added

zaynt4606 reviewed (commented):
plz run `dev/reformat` for [Style check] 
and `UPDATE=1 build/mvn clean test -pl common -am -Dtest=none -DwildcardSuites=org.apache.celeborn.ConfigurationSuite` for configuration change~

s0nskar reviewed (commented):
@saurabhd336 there seems to be some issues with the tests.

SteNicholas commented:
@saurabhd336, please create new JIRA ticket for this pull request.

saurabhd336 commented:
@zaynt4606 @s0nskar @SteNicholas Thanks for helping review this. I've fixed the tests + lint issues.
@SteNicholas could you please help me how to create a celeborn ticket for this change?

SteNicholas commented:
@saurabhd336, you could refer to #1053 for creating new JIRA tiket.

saurabhd336 commented on worker/src/test/scala/org/apache/celeborn/service/deploy/cluster/ReadWriteTestWithFailures.scala:
```diff
@@ -0,0 +1,112 @@
+    shutdownMiniCluster()
+  }
+
+  test(s"test MiniCluster with connection resets, ensure no duplicate reads") {
```
Added.

saurabhd336 commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -4691,6 +4692,15 @@ object CelebornConf extends Logging {
+  val WORKER_PARTITION_READER_CHECKPOINT_ENABLE: ConfigEntry[Boolean] =
+    buildConf("celeborn.worker.partition.reader.checkpointEnabled")
+      .categories("client")
+      .version("0.5.0")
```
Ack

zaynt4606 commented on client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java:
```diff
@@ -210,14 +224,34 @@ public PartitionLocation getLocation() {
+                  + " likely by a previous reader for the same partition.",
+              chunkIndex);
+          chunkIndex++;
+          returnedChunks++;
```
why `toFetch` dont need to decrease if `toFetch` is `fetchMaxReqsInFlight - inFlight + 1` when `chunkIdsAlreadyReturned.contains(chunkIndex)`

saurabhd336 replied:
The way I read `toFetch`, it seems like it's trying to ensure no more than `fetchMaxReqsInFlight` requests are submitted at once, while ensuring we fetch as many chunks as possible at once.

My thought process is that if we're skipping certain chunks, we could instead fetch other chunks in the list. WDYT?

saurabhd336 replied:
Actually, come to think of it, incrementing toFetch here would be wrong and cause an infinite wait. I added a comment explaining it.

zaynt4606 replied:
Got it.
The `returnedChunks` increased and the chunk actually to fetch has no change.

zaynt4606 commented on client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java:
```diff
@@ -185,7 +196,10 @@ public void close() {
-      results.forEach(ReferenceCounted::release);
+      results.forEach(
+          chunk -> {
+            chunk.getRight().release(); //
```
empty annotation

saurabhd336 replied:
Fixed.

saurabhd336 commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -982,6 +982,8 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
   def clientFetchBufferSize: Int = get(CLIENT_FETCH_BUFFER_SIZE).toInt
   def clientFetchMaxReqsInFlight: Int = get(CLIENT_FETCH_MAX_REQS_IN_FLIGHT)
+  def isWorkerPartitionReaderCheckpointEnabled: Boolean =
+    get(WORKER_PARTITION_READER_CHECKPOINT_ENABLE)
```
Makes sense. Ack

saurabhd336 replied:
@zaynt4606 It is possible to fallback to reading from the same server even when replication is enabled. Eg: when `celeborn.client.adaptive.optimizeSkewedPartitionRead.enabled` is set to true and partitions are being split. I'd let `CelebornInputStream` itself decide whether or not to restore checkpoint when creating a reader

SteNicholas commented on client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java:
```diff
@@ -417,14 +418,27 @@ private boolean isExcluded(PartitionLocation location) {
+        Optional<PartitionReaderCheckpointMetadata> checkpointMetadata)
+        throws IOException {
       Exception lastException = null;
+      PartitionReader reader = null;
```
Could move this definition into line 437? The catch blocker does not use this variable.

saurabhd336 replied:
Ack

SteNicholas commented on client/src/main/java/org/apache/celeborn/client/read/PartitionReader.java:
```diff
@@ -21,14 +21,21 @@
 
   PartitionLocation getLocation();
+
+  default T getPartitionReaderCheckpointMetadata() {
```
Why does this method has default implementation?

saurabhd336 replied:
I initially thought of implementing it only for the WorkerPartitionReader, but I can add no-op impl for other types of readers, and can take up their implementation in a subsequent PR

SteNicholas commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -4691,6 +4698,15 @@ object CelebornConf extends Logging {
       .createWithDefaultString("64k")
 
+  val WORKER_PARTITION_READER_CHECKPOINT_ENABLE: ConfigEntry[Boolean] =
+    buildConf("celeborn.worker.partition.reader.checkpointEnabled")
```
```suggestion
    buildConf("celeborn.worker.partition.reader.checkpoint.enabled")
```

saurabhd336 replied:
Ack

SteNicholas commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -4691,6 +4698,15 @@ object CelebornConf extends Logging {
       .bytesConf(ByteUnit.BYTE)
       .createWithDefaultString("64k")
 
+  val WORKER_PARTITION_READER_CHECKPOINT_ENABLE: ConfigEntry[Boolean] =
```
```suggestion
  val WORKER_PARTITION_READER_CHECKPOINT_ENABLED: ConfigEntry[Boolean] =
```

saurabhd336 replied:
Ack

FMX commented on client/src/main/java/org/apache/celeborn/client/read/checkpoint/WorkerPartitionReaderCheckpointMetadata.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Set;
+
+/** Checkpoint metadata for a partition reader on the worker side. */
+public class WorkerPartitionReaderCheckpointMetadata implements PartitionReaderCheckpointMetadata {
```
This class can be used in worker partition readers, DFS partition reader, and local partition readers. IMO, here is no need to extract an empty interface for it.

saurabhd336 replied:
I initially thought of implementing it only for the WorkerPartitionReader, but I can add no-op impl for other types of readers, and can take up their implementation in a subsequent PR

FMX commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -982,6 +982,13 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
+    if (clientPushReplicateEnabled) {
+      false
+    } else
+      get(WORKER_PARTITION_READER_CHECKPOINT_ENABLE)
```
Save it to a val to avoid repeatedly parsing this config

saurabhd336 commented:
@zaynt4606 @FMX @SteNicholas @s0nskar Review comments have been addressed PTAL!

FMX reviewed (commented):
Looks like this optimization is generic for all partition readers.

saurabhd336 commented on client/src/main/java/org/apache/celeborn/client/read/DfsPartitionReader.java:
```diff
@@ -308,4 +310,15 @@ private void closeStream() {
+
+  @Override
+  public Optional<PartitionReaderCheckpointMetadata> getPartitionReaderCheckpointMetadata() {
+    // TODO implement similar to {@link WorkerPartitionReader}
```
I had planned to implement in a followup PR, after testing this for worker partition reader. Anway, I've added the implementation for DfsPartitionReader. For LocalPartitionReader, I think the complexity of implementation outweighs the potential benefits (since for both the worker / dfs readers, its a network call to fetch the chunks, while for local reader it's just a local file buffer read)

I can take up the impl for local reader in a followup PR. wdyt?

FMX commented on client/src/main/java/org/apache/celeborn/client/read/DfsPartitionReader.java:
```diff
@@ -238,7 +257,7 @@ public ByteBuf next() throws IOException, InterruptedException {
                   }
                 }
-                results.put(Unpooled.wrappedBuffer(buffer));
+                results.put(Pair.of(currentChunkIndex, Unpooled.wrappedBuffer(buffer)));
```
A chunk that has been returned does not mean that the chunk is read fully. IMO, I think you should return the chunk index with the chunk and record the chunk index as fully read after the celeborn input stream consumed the chunk fully.

FMX commented on client/src/main/java/org/apache/celeborn/client/read/PartitionReader.java:
```diff
@@ -31,4 +33,8 @@ public interface PartitionReader {
 
   PartitionLocation getLocation();
+
+  Optional<PartitionReaderCheckpointMetadata> getPartitionReaderCheckpointMetadata();
```
You can add a chunk index in the method of `next` to make sure that you can record the chunk that has been fully consumed.

RexXiong commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -4691,6 +4694,15 @@ object CelebornConf extends Logging {
       .createWithDefaultString("64k")
 
+  val PARTITION_READER_CHECKPOINT_ENABLED: ConfigEntry[Boolean] =
+    buildConf("celeborn.partition.reader.checkpoint.enabled")
```
Maybe rename the configuration to `celeborn.client.partition.reader.checkpoint.enabled`, since it is only for client use.

saurabhd336 replied:
Ack

RexXiong commented on client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java:
```diff
@@ -176,7 +186,8 @@ public ByteBuf next() throws IOException, InterruptedException {
     }
     returnedChunks++;
-    return chunk;
+    chunkIdsAlreadyReturned.add(chunk.getLeft());
```
Only checkpoint when isCheckpointEnabled

saurabhd336 replied:
With the new way of checkpointing, this is now handled

RexXiong commented on client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java:
```diff
@@ -210,14 +224,39 @@ public PartitionLocation getLocation() {
+      int toFetch = Math.min(fetchMaxReqsInFlight - inFlight + 1, endChunkIndex + 1 - chunkIndex);
+
+      while (toFetch > 0 && chunkIndex <= endChunkIndex) {
+        if (chunkIdsAlreadyReturned.contains(chunkIndex)) {
```
`isCheckpointEnabled && chunkIdsAlreadyReturned.contains(chunkIndex)`

saurabhd336 replied:
Now handled

FMX commented:
> @FMX I noticed the comment regarding checkpointing only after a chunk is fully read, while i'm not sure under what cases can a returned chunk be not fully read before calling `next()`, I nonetheless have addressed the comment, now checkpointing last returned chunk id on each `next()` call. PTAL!

Hi, after some discussion, we found that if a chunk is returned and is not fully read, the spark task will fail, so this won't be a problem here.

saurabhd336 commented:
> > @FMX I noticed the comment regarding checkpointing only after a chunk is fully read, while i'm not sure under what cases can a returned chunk be not fully read before calling `next()`, I nonetheless have addressed the comment, now checkpointing last returned chunk id on each `next()` call. PTAL!
> 
> Hi, after some discussion, we found that if a chunk is returned and is not fully read, the spark task will fail, so this won't be a problem here.

Yes that's what I had noticed too, I've anyway changes the pr in a way that the last chunk returned is now checkpointed in the next `next()` call, ensuring we only checkpoint once the last returned chunk is fully read. PTAL!

RexXiong commented on client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java:
```diff
@@ -156,13 +173,21 @@ public boolean hasNext() {
   @Override
   public ByteBuf next() throws IOException, InterruptedException {
     checkException();
+    checkpoint();
```
nit: IMO `checkpoint` can be placed before `checkException`

saurabhd336 replied:
Yes makes sense. Ack

RexXiong reviewed (approved):
LGTM, only a nit

saurabhd336 commented:
@zaynt4606  @FMX @SteNicholas  @s0nskar  PTAL!

RexXiong commented:
Thanks. merge to main(v0.6.0)
